src/gpio.rs

```diff
@@ -502,6 +518,43 @@ macro_rules! gpio {
                         unsafe { (*$GPIOX::ptr()).idr.read().bits() & (1 << $i) == 0 }
                     }
                 }
+
+                impl<MODE> ExtiPin for $PXi<Input<MODE>> {
+                    /// generate interrupt on rising edge, falling edge or both
+                    fn trigger_on_edge(&mut self, exti: &mut EXTI, edge: Edge) {
+                        match edge {
+                            Edge::RISING => {
+                                exti.rtsr.write(|w| { w.$tri().set_bit()   });
+                                exti.ftsr.write(|w| { w.$tri().clear_bit() });
+                            },
+                            Edge::FALLING => {
+                                exti.ftsr.write(|w| { w.$tri().set_bit()   });
+                                exti.rtsr.write(|w| { w.$tri().clear_bit() });
+                            },
+                            Edge::RISING_FALLING => {
+                                exti.rtsr.write(|w| { w.$tri().set_bit() });
+                                exti.ftsr.write(|w| { w.$tri().set_bit() });
+                            }
+                        }
+                    }
+
+                    /// Enable external interrupts from this pin.
+                    /// Note: This configures EXTI line $i to trigger from this pin
+                    fn enable_interrupt(&mut self, exti: &mut EXTI, syscfg: &mut SYSCFG) {
```

Comment:
I'm feeling like it might be better to not need syscfg as an argument here, and instead do the same as we do for RCC everywhere :
```
let syscfg = unsafe { &(*SYSCFG::ptr()) };
```
Note that I have not really thought about all the implications it may have, so I may be wrong about this.
Also note that the same remark might also apply to the exti argument, but I'm even less sure about it.

Reply:
Actually we might have been a bit too lax here in the past. For RMW operations we should always require either a `&mut` on the resource or a `CriticialSection` , potentially we could also offer both options as a convenience.

Reply:
You mean a safe function that requires the SYSCFG and an unsafe one which works like @aurelj suggested?

Reply:
True. What do you think about introducing another function handling only the `SYSCFG` part of the configuration? This would separate interrupt configuration from enabling and disabling them. The new function then only needs access to `SYSCFG` leaving only `EXTI` for `enable_interrupt()`.

Reply:
@arkap Sounds fantastic.